File: src/parser/ast.rs
use crate::{span::Span, vm::register::Register};

pub use crate::parser::immediate::*;

#[derive(Debug, Clone, PartialEq)]
pub enum Statement {
    Label(String, Span),

    Error(Expression, Span),
    Define(Expression, Expression, Span),
    Include(Expression, Span),
    IfDef(Expression, Span),
    IfNDef(Expression, Span),
    Else(Span),
    EndIf(Span),

    Section(SectionType, Span),
    Entry(Expression, Span),
    Ascii(Expression, Span),
    Asciz(Expression, Span),

    Nop(Span),
    Mov(Expression, Expression, Span),
    Ldr(Expression, Expression, Span),
    Str(Expression, Expression, Span),
    Push(Option<Expression>, Expression, Span),
    Pop(Option<Expression>, Expression, Span),
    Add(Expression, Expression, Expression, Span),
    Sub(Expression, Expression, Expression, Span),
    Mul(Expression, Expression, Expression, Span),
    Div(Expression, Expression, Expression, Span),
    And(Expression, Expression, Expression, Span),
    Or(Expression, Expression, Expression, Span),
    Xor(Expression, Expression, Expression, Span),
    Shl(Expression, Expression, Expression, Span),
    Shr(Expression, Expression, Expression, Span),
    Cmp(Expression, Expression, Span),
    Jmp(Expression, Span),
    Jne(Expression, Span),
    Jeq(Expression, Span),
    Jlt(Expression, Span),
    Jgt(Expression, Span),
    Jle(Expression, Span),
    Jge(Expression, Span),
    Call(Expression, Span),
    Ret(Span),
    Inc(Expression, Span),
    Dec(Expression, Span),
    Syscall(Span),
    Hlt(Span),

    Db(Vec<Expression>, Span),
    Resb(Expression, Span),
}

impl Statement {
    pub fn span(&self) -> Span {
        match self {
            Statement::Error(_, span) => *span,
            Statement::Define(_, _, span) => *span,
            Statement::Include(_, span) => *span,
            Statement::IfDef(_, span) => *span,
            Statement::IfNDef(_, span) => *span,
            Statement::Else(span) => *span,
            Statement::EndIf(span) => *span,
            Statement::Section(_, span) => *span,
            Statement::Entry(_, span) => *span,
            Statement::Ascii(_, span) => *span,
            Statement::Asciz(_, span) => *span,
            Statement::Label(_, span) => *span,
            Statement::Nop(span) => *span,
            Statement::Mov(_, _, span) => *span,
            Statement::Ldr(_, _, span) => *span,
            Statement::Str(_, _, span) => *span,
            Statement::Push(_, _, span) => *span,
            Statement::Pop(_, _, span) => *span,
            Statement::Add(_, _, _, span) => *span,
            Statement::Sub(_, _, _, span) => *span,
            Statement::Mul(_, _, _, span) => *span,
            Statement::Div(_, _, _, span) => *span,
            Statement::And(_, _, _, span) => *span,
            Statement::Or(_, _, _, span) => *span,
            Statement::Xor(_, _, _, span) => *span,
            Statement::Shl(_, _, _, span) => *span,
            Statement::Shr(_, _, _, span) => *span,
            Statement::Cmp(_, _, span) => *span,
            Statement::Jmp(_, span) => *span,
            Statement::Jne(_, span) => *span,
            Statement::Jeq(_, span) => *span,
            Statement::Jlt(_, span) => *span,
            Statement::Jgt(_, span) => *span,
            Statement::Jle(_, span) => *span,
            Statement::Jge(_, span) => *span,
            Statement::Call(_, span) => *span,
            Statement::Ret(span) => *span,
            Statement::Inc(_, span) => *span,
            Statement::Dec(_, span) => *span,
            Statement::Syscall(span) => *span,
            Statement::Hlt(span) => *span,
            Statement::Db(_, span) => *span,
            Statement::Resb(_, span) => *span,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expression {
    Identifier(String),
    Register(Register),
    IntegerLiteral(i64),
    FloatLiteral(f64),
    StringLiteral(String),
    DataSize(DataSize),
    Address(Box<Expression>, Option<Box<Expression>>),
    BinaryOp(Box<Expression>, BinaryOperator, Box<Expression>, Span),
}

#[derive(Debug, Clone, PartialEq)]
pub enum SectionType {
    Text,
    Data,
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinaryOperator {
    Add,    // +
    Sub,    // -
    Mul,    // *
    Div,    // /
    BitOr,  // |
    BitAnd, // &
    BitXor, // ^
}

File: src/parser/ast.zig
// TODO: all expressions have a span for better error reporting

const std = @import("std");
const Span = @import("../Span.zig");
const StringInterner = @import("../StringInterner.zig");
const StringId = StringInterner.StringId;
const DataSize = @import("immediate.zig").DataSize;
const Register = @import("../vm/register.zig").Register;

pub const Statement = union(enum) {
    label: Label,
    @"error": Expr1,
    define: Define,
    include: Expr1,
    ifdef: Expr1,
    ifndef: Expr1,
    @"else": Span,
    endif: Span,
    section: Section,
    entry: Expr1,
    ascii: Expr1,
    asciz: Expr1,
    @"extern": Expr1,
    nop: Span,
    mov: Expr2,
    ldr: Expr2,
    str: Expr2,
    sti: Expr3,
    push: PushPop,
    pop: PushPop,
    add: Expr3,
    sub: Expr3,
    mul: Expr3,
    div: Expr3,
    @"and": Expr3,
    @"or": Expr3,
    xor: Expr3,
    shl: Expr3,
    shr: Expr3,
    cmp: Expr2,
    jmp: Expr1,
    jne: Expr1,
    jeq: Expr1,
    jlt: Expr1,
    jgt: Expr1,
    jle: Expr1,
    jge: Expr1,
    call: Expr1,
    ret: Span,
    inc: Expr1,
    dec: Expr1,
    neg: Expr1,
    syscall: Span,
    hlt: Span,
    db: Db,
    dw: Db,
    dd: Db,
    dq: Db,
    resb: Expr1,
    resw: Expr1,
    resd: Expr1,
    resq: Expr1,

    pub const Expr1 = struct {
        expr: *Expression,
        span: Span,
    };

    pub const Expr2 = struct {
        expr1: *Expression,
        expr2: *Expression,
        span: Span,
    };

    pub const Expr3 = struct {
        expr1: *Expression,
        expr2: *Expression,
        expr3: *Expression,
        span: Span,
    };

    pub const Label = struct {
        name: StringId,
        span: Span,
    };

    pub const Define = struct {
        name: *Expression,
        expr: ?*Expression,
        span: Span,
    };

    pub const Section = struct {
        type: Type,
        span: Span,
        pub const Type = enum { text, data };
    };

    pub const PushPop = struct {
        data_size: ?*Expression,
        expr: *Expression,
        span: Span,
    };

    // TODO: each expr should have its own span
    pub const Db = struct {
        exprs: []*Expression,
        span: Span,
    };

    pub fn span(self: Statement) Span {
        return switch (self) {
            .label => |v| v.span,
            .@"error" => |v| v.span,
            .define => |v| v.span,
            .include => |v| v.span,
            .ifdef => |v| v.span,
            .ifndef => |v| v.span,
            .@"else" => |v| v,
            .endif => |v| v,
            .section => |v| v.span,
            .entry => |v| v.span,
            .ascii => |v| v.span,
            .asciz => |v| v.span,
            .@"extern" => |v| v.span,
            .nop => |v| v,
            .mov => |v| v.span,
            .ldr => |v| v.span,
            .str => |v| v.span,
            .sti => |v| v.span,
            .push => |v| v.span,
            .pop => |v| v.span,
            .add => |v| v.span,
            .sub => |v| v.span,
            .mul => |v| v.span,
            .div => |v| v.span,
            .@"and" => |v| v.span,
            .@"or" => |v| v.span,
            .xor => |v| v.span,
            .shl => |v| v.span,
            .shr => |v| v.span,
            .cmp => |v| v.span,
            .jmp => |v| v.span,
            .jne => |v| v.span,
            .jeq => |v| v.span,
            .jlt => |v| v.span,
            .jgt => |v| v.span,
            .jle => |v| v.span,
            .jge => |v| v.span,
            .call => |v| v.span,
            .ret => |v| v,
            .inc => |v| v.span,
            .dec => |v| v.span,
            .neg => |v| v.span,
            .syscall => |v| v,
            .hlt => |v| v,
            .db => |v| v.span,
            .dw => |v| v.span,
            .dd => |v| v.span,
            .dq => |v| v.span,
            .resb => |v| v.span,
            .resw => |v| v.span,
            .resd => |v| v.span,
            .resq => |v| v.span,
        };
    }
};

pub const Expression = union(enum) {
    identifier: StringId,
    register: Register,
    integer_literal: i64,
    float_literal: f64,
    string_literal: StringId,
    data_size: DataSize,
    address: Address,
    unary_op: UnaryOp,
    binary_op: BinaryOp,

    pub const Address = struct {
        base: *Expression,
        offset: ?*Expression,
    };

    pub const UnaryOp = struct {
        expr: *Expression,
        op: Op,
        span: Span,

        pub const Op = enum {
            neg,
        };
    };

    pub const BinaryOp = struct {
        lhs: *Expression,
        op: Op,
        rhs: *Expression,
        span: Span,

        pub const Op = enum {
            add, // +
            sub, // -
            mul, // *
            div, // /
            bit_or, // |
            bit_and, // &
            bit_xor, // ^
        };
    };
};
